### ISSUE TITLE
upgrade recommendation for existing user

### ISSUE THREAD

**Issue by Hugo2121** [Quang-Vinh/matrix-factorization#4]
```
        for user in known_users:
            user_index = self.user_id_map[user]

            # Initialize bias
            self.user_biases[user_index] = 0

            # Initialize latent factors vector
            self.user_features[user_index, :] = np.random.normal(
                self.init_mean, self.init_sd, (1, self.n_factors)
            )

```

why we need to re-initialize params for old users? old user has a meaningful user matrix already, why don't we just continue training on the user matrix?

for example, my use case does not limit the max rating to 5 (the actual rating could be over 1000). 

for user 0:
if I have an item A with rating 1000, this item should have most popular rating, and get recommended. 

then I tried to update the item B with rating to 500, since user matrix is re-initialized, the result is wield, it does not recommend item A, although item A has a rating 1000.

if I comment out the above code, the top 10 recommendation result will show item A and item B, which by my understanding, should be the correct recommendation.

**Comment by Quang-Vinh** [comment]
That is a good point, however the order in which ratings are visited in SGD is important as described in the paper and so it suggests retraining the entire user feature vector. 

It seems odd that the item with rating 1000 is not being recommended. Are you providing the entire user ratings list when retraining or just only providing the new ratings? If possible can you provide some example code to reproduce your problem?

**Comment by Hugo2121** [comment]
thanks for your reply.

the code is as follows:

**1. after training the model with sample data, I get the original recommendation for user 200:**

![image](https://user-images.githubusercontent.com/29109034/108754581-f5d5af00-74fa-11eb-9d33-fda16ec40111.png)


**2. then I update the user 200 on item 16 with rating 100**
```
new_user_data = {'user_id': [200], 'item_id': [16]}
new_user_rating = [100]

new_data = pd.DataFrame(new_user_data)
new_rating = pd.DataFrame(new_user_rating)

# Update model with new users
matrix_fact.update_users(
    new_data, new_rating, lr=0.01, n_epochs=2000, verbose=1
)
```

**3. check the recommend result for user 200, I can see item 16, this seems correct**

![image](https://user-images.githubusercontent.com/29109034/108755028-92984c80-74fb-11eb-9907-a17d86f7f606.png)

**4. then I tried to update user with item 20, the recommend result suppose to have both 16 and 20, but the result is:**
```
new_user_data = {'user_id': [200], 'item_id': [20]}
new_user_rating = [100]

new_data = pd.DataFrame(new_user_data)
new_rating = pd.DataFrame(new_user_rating)

# Update model with new users
matrix_fact.update_users(
    new_data, new_rating, lr=0.01, n_epochs=2000, verbose=1
)

```
 
![image](https://user-images.githubusercontent.com/29109034/108755093-a8a60d00-74fb-11eb-93c6-add60cc60a51.png)




so item 16 is not in top 10 recommended result, which seems not correct to me

**Comment by Quang-Vinh** [comment]
Thank you for the code samples. 

When calling matrix_fact.update_users() you should pass the entire new user rating list, so for your second update in part 4 your matrix should be

```
new_user_data = {'user_id': [200, 200], 'item_id': [16, 20]}
new_user_rating = [100, 100]
```

The reason it is like this is to allow for updated ratings to the user rating list, for example if user 200 rates item 16 with rating 200 but then later changes the rating to 100. And also to allow deleting items in a user's list which would mean we want to re-initialize their feature vector and perform SGD again. 

Let me know if that makes sense.

**Comment by Hugo2121** [comment]
Thanks, I get your idea. 

1. Do you recommend update user with all data or only newly generate data? my understanding is only update with entire newly generate data ( do not include original data for user 200 which is unchanged)
2. how do I know the model is not overfitted? just try some unseen user and check validation error?

Thanks

**Comment by Quang-Vinh** [comment]
1. You would need to update with all original data for user 200. Or another way of thinking of it is that you would have to provide their entire current user rating list which could contain new, dropped, updated or not updated ratings.

2. In general for supervised learning to check overfitting you should apply some validation techniques. Either by using train/val split or by applying cross-validation. Though you would always want to reserve a test set. The package attempts to be compatible with some scikit-learn methods so I would recommend directly using those methods as well as shown in the examples notebook at the very end.

**Comment by Hugo2121** [comment]
thanks for your help
